accepted/2023/net8.0-browser-tfm.md

```diff
@@ -0,0 +1,187 @@
+# net8.0-browser TFM for applications running in the browser
+
+**Owner** [Javier Calvarro](https://github.com/javiercn) | [Javier Calvarro](https://github.com/javiercn)
+**Contact** [Daniel Roth](https://github.com/danroth27) | [Daniel Roth](https://github.com/danroth27)
+**Contact** [Artak Mkrtchyan](https://github.com/mkartakmsft) | [Artak Mkrtchyan](https://github.com/mkartakmsft)
+
+For a video introduction to what we are trying to achieve, see:
+https://www.youtube.com/live/kIH_Py8ytlY?feature=share&t=473
+https://www.youtube.com/watch?v=48G_CEGXZZM
+
+We have been shipping .NET in the browser via webassembly since the introduction of Blazor 3.2.0. During that initial release we decided to avoid adding a TFM for the browser as we were not clear on whether we needed it and we knew it was a one way operation. Once we introduced a TFM, we would never be able to take it back.
+
+Our reasoning was that we could likely get away with a RID as that would allow third-parties to provide alternative pre-compiled webassembly assets (including native dependencies compiled to wasm).
+
+We reasoned that we did not need a TFM as we could always provide a PAL over any functionality we needed and we would annotate the APIs with the `[SupportedOSPlatform]` attribute and an analyzer that would warn against incorrect usage when targeting the browser.
+
+Over time, we have learned that the lack of a TFM is limiting for us and our customers. The lack of a TFM introduces friction in two ways:
+* **Exposing additional customer-facing APIs when running in browser (and not on desktop)**: For example, Blazor includes many platform agnostic abstractions like IJSRuntime, IJSInProcessRuntime and different implementations to be able to reflect the different capabilities of the different platforms. This makes taking advantage of some of the webassembly unique capabilities more challenging as this kind of functionality needs to be carefully designed, and creates other problems like lack of discoverability (you need to know the existence of IJSInProcessRuntime and downcast to it).
+  ```mermaid
+  classDiagram
+    IJSRuntime <|-- IJSInProcessRuntime
+    class IJSRuntime {
+      <<interface>>
+      InvokeAsync<&zwj;TValue>(string identifier, object[] args) ValueTask~TValue~
+      InvokeAsync<&zwj;TValue>(string identifier, CancellationToken cancellationToken, object?[]? args) ValueTask~TValue~
+    }
+    class IJSInProcessRuntime {
+      <<interface>>
+      Invoke<&zwj;TResult>(string identifier, params object?[]? args) TResult
+    }
+  ```
+* **Varying framework/library code contents/behavior based on whether it's running in the browser or not:** For example, having DI dependencies or other library code references varying by platform. Doing it based on TFM (not on RuntimeInformation) helps compilation and trimming.
```

Comment:
I thought our linker is able to tell that the return values for `OSPlatform.IsBrowser()` is a constant. Is that not true?

Reply:
In certain *very carefully balanced* cases it's true that the linker may resolve `OSPlatform.IsBrowser()`, discard some `if`/`else` branches, and consequently discard entire assembly references. That's great if it works out.

However in practice it's really difficult to make this happen reliably. If you assign `OSPlatform.IsBrowser()` to a variable then check based on that, or if you have some other field referencing a type that's only in one of these dependencies, or if you simply try to publish an app without trimming (which app developers are always within their rights to do) then this level of reliance on the linker will get you into trouble.

If we were going to rely on this, we would need better infrastructure to assert at publish time that certain assemblies must not end up in the output or that only a specified set can do so.

And even then, it wouldn't help guide application developers to avoid referencing types that aren't usable on their target platform. For example, we may reference EF conditionally based on your TFM, and we want developers to know at compile time that they can only use those EF APIs in the `net8.0` build and not in the `net8.0-browser` build.

Reply:
> In certain very carefully balanced cases it's true that the linker may resolve OSPlatform.IsBrowser(), discard some if/else branches, and consequently discard entire assembly references. That's great if it works out.

Not only that but linker, JIT, AOT compiler and also many of our C# analyzers can do that as well (making the code effectively disabled during coding) so using ifdefs is not really giving you any perf/size benefits.

Reply:
The claim I'm making is that relying on what symbols happen to be statically referenced is too fragile in reality. There's nothing stopping a developer from accidentally/ignorantly referencing symbols from one of the assemblies that you really wanted to get linked out. Compiler enforcement is required, which is what the TFM enables.

Reply:
I assumed our linker is smart enough, but if you find that this is too fragile do deliver what you need, then I'll take your word for it.

Reply:
In the discussion it became clear that this isn't about the linker; it's that the developer wants a single application project with the ability to depend on different components between client and server while also having the tools that avoids accidentally pulling server component to the client, which is what multi-targeting would solve, but it requires a TFM.